File: apps/web/lib/api/groups.ts
import type { paths } from "@ims/api-types"
import { apiFetch } from "./client"

export type GroupSummary = paths["/api/groups"]["get"]["responses"]["200"]["content"]["application/json"][number]
export type GroupDetail = paths["/api/groups/{group_id}"]["get"]["responses"]["200"]["content"]["application/json"]
export type GroupMember = GroupDetail["members"][number]
export type GroupCreateBody = paths["/api/groups"]["post"]["requestBody"]["content"]["application/json"]
export type GroupUpdateBody = paths["/api/groups/{group_id}"]["patch"]["requestBody"]["content"]["application/json"]

export async function listGroups(token: string | null): Promise<GroupSummary[]> {
  return (await apiFetch("/groups", { accessToken: token })).json()
}
export async function createGroup(body: GroupCreateBody, token: string | null): Promise<GroupSummary> {
  return (await apiFetch("/groups", { method: "POST", accessToken: token, body: JSON.stringify(body) })).json()
}
export async function getGroup(id: string, token: string | null): Promise<GroupDetail> {
  return (await apiFetch(`/groups/${id}`, { accessToken: token })).json()
}
export async function updateGroup(id: string, body: GroupUpdateBody, token: string | null): Promise<GroupSummary> {
  return (await apiFetch(`/groups/${id}`, { method: "PATCH", accessToken: token, body: JSON.stringify(body) })).json()
}
export async function deleteGroup(id: string, token: string | null): Promise<void> {
  await apiFetch(`/groups/${id}`, { method: "DELETE", accessToken: token })
}
export async function addGroupMember(groupId: string, username: string, token: string | null): Promise<GroupMember> {
  return (await apiFetch(`/groups/${groupId}/members`, { method: "POST", accessToken: token, body: JSON.stringify({ username }) })).json()
}
export async function removeGroupMember(groupId: string, userId: string, token: string | null): Promise<void> {
  await apiFetch(`/groups/${groupId}/members/${userId}`, { method: "DELETE", accessToken: token })
}

File: apps/api/app/routes/groups.py
from __future__ import annotations
from uuid import UUID

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.group import (
    GroupAddMember,
    GroupCreate,
    GroupDetail,
    GroupMemberRead,
    GroupSummary,
    GroupUpdate,
)
from app.services import groups_service as svc

router = APIRouter(prefix="/api/groups", tags=["groups"])


@router.get("", response_model=list[GroupSummary])
async def list_groups(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    return await svc.list_groups(session, user.id)


@router.post("", response_model=GroupSummary, status_code=status.HTTP_201_CREATED)
async def create_group(
    body: GroupCreate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    return await svc.create_group(session, user.id, body)


@router.get("/{group_id}", response_model=GroupDetail)
async def get_group(
    group_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    return await svc.get_group_detail(session, user.id, group_id)


@router.patch("/{group_id}", response_model=GroupSummary)
async def update_group(
    group_id: UUID,
    body: GroupUpdate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    return await svc.update_group(session, user.id, group_id, body)


@router.delete(
    "/{group_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def delete_group(
    group_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    await svc.delete_group(session, user.id, group_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/{group_id}/members", response_model=GroupMemberRead, status_code=status.HTTP_201_CREATED
)
async def add_member(
    group_id: UUID,
    body: GroupAddMember,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    return await svc.add_member_by_username(session, user.id, group_id, body)


@router.delete(
    "/{group_id}/members/{user_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def remove_member(
    group_id: UUID,
    user_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
):
    await svc.remove_member(session, user.id, group_id, user_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
